### ISSUE TITLE
Cacheing of variables in foreach fails

### ISSUE THREAD

**Issue by 0xFEEDC0DE64** [smarty-php/smarty#110]
I need to cache a big list in a foreach. But I dont want to cache which one will be highlighted.
The {foreach} itself gets cached, but not variable in the {if}!

See this very small test case:

testcase.php:
```
<?php

require_once('smarty/Smarty.class.php');

$smarty = new Smarty;

$smarty->setCaching(true);

if(!$smarty->isCached('testcase.tpl')) {
    $smarty->assign('list', [
        ['id' => 1, 'name' => 'test 1'],
        ['id' => 2, 'name' => 'test 2'],
        ['id' => 3, 'name' => 'test 3']
    ]);
}

$smarty->assign('current', 2, true);

$smarty->display('testcase.tpl');
```
testcase.tpl
```
<ul>
{foreach from=$list item=item}
    <li>{$item.name} {if $current==$item.id} ACTIVE{/if}</li>
{/foreach}
</ul>
```

It is not possible because the if accesses the variable $item even if cache is enabled!
```if ($_smarty_tpl->tpl_vars[\'current\']->value == $_smarty_tpl->tpl_vars[\'item\']->value[\'id\'])```

it should look like this (note that this is handmade, not compilermade):

```if ($_smarty_tpl->tpl_vars[\'current\']->value == 12)```

[testcase.php](https://github.com/smarty-php/smarty/files/47205/testcase.php.txt)
[testcase.tpl](https://github.com/smarty-php/smarty/files/47204/testcase.tpl.txt)
[7af6b444214920055f0e476b62736e9e88058c99_0.file.testcase.tpl.cache.php](https://github.com/smarty-php/smarty/files/47207/7af6b444214920055f0e476b62736e9e88058c99_0.file.testcase.tpl.cache.php.txt)
[7af6b444214920055f0e476b62736e9e88058c99.testcase.tpl.php](https://github.com/smarty-php/smarty/files/47206/7af6b444214920055f0e476b62736e9e88058c99.testcase.tpl.php.txt)

```

**Comment by uwetews** [comment]
It's a know restriction that you can't combine normal variables and nocache variables in one tag.
We have this on our todo list, but first experiments showed that it is a very complex task.

Sorry I can't help at the moment.

**Comment by 0xFEEDC0DE64** [comment]
Isn't it possible that you check every expression while generating the compiled template if it is a cached expression?

**Comment by uwetews** [comment]
The problem is that the value of $item.id is not know at  compile time. An even worse the value of $item.id does change on each iteration of the foreach loop. The foreach loop is processed while rendering the compiled template. and the variable $item.id does exist only at that time. The cached template does not know any thing about all of this.

**Comment by uwetews** [comment]
There is now a solution for your problem with the new tag {make_nocache}.
Read NEW_FEATURES.txt

```php
<ul>
{foreach from=$list item=item}
    <li>{$item.name} {make_nocache $item}}{if $current==$item.id} ACTIVE{/if}</li>
{/foreach}
</ul>
```
This is now available in the master branch and will later be included in 3.1.30
